File: codes/matlab/ch04/lebesgue_functions.m
%% lebesgue_functions.m
%
% Visualizes and compares Lebesgue functions for different node distributions.
%
% The Lebesgue function is Lambda_N(x) = sum |L_k(x)|, and its maximum
% is the Lebesgue constant Lambda_N.
%
% Asymptotic growth rates:
%   - Chebyshev: Lambda_N ~ (2/pi) ln(N)
%   - Legendre:  Lambda_N ~ O(sqrt(N))
%   - Equispaced: Lambda_N ~ 2^N / (N ln N)
%
% Author: Dr. Denys Dutykh
%         Mathematics Department
%         Khalifa University of Science and Technology
%         Abu Dhabi, UAE
%
% Part of the book "Computational Études: A Spectral Approach"

clear; close all; clc;

%% Configuration
N_FINE = 1000;

% Book color scheme
NAVY = [20, 45, 110] / 255;
SKY = [120, 150, 210] / 255;
CORAL = [231, 76, 60] / 255;
TEAL = [22, 160, 133] / 255;

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch04', 'matlab');
output_file = fullfile(output_dir, 'lebesgue_functions.pdf');

%% Fine grid
x_fine = linspace(-1, 1, N_FINE)';

%% Create figure
fig = figure('Units', 'inches', 'Position', [1, 1, 11, 4.5]);

% Left panel: Lebesgue functions for N = 10
subplot(1, 2, 1);
hold on; box on;

N = 10;

% Equispaced
x_equi = linspace(-1, 1, N+1)';
Lambda_equi = lebesgue_function(x_equi, x_fine);

% Chebyshev
j = 0:N;
x_cheb = cos(j * pi / N)';
Lambda_cheb = lebesgue_function(x_cheb, x_fine);

% Legendre-Gauss-Lobatto (approximation using Legendre polynomial roots)
x_leg = legendre_nodes(N);
Lambda_leg = lebesgue_function(x_leg, x_fine);

plot(x_fine, Lambda_equi, 'Color', CORAL, 'LineWidth', 1.5, ...
     'DisplayName', sprintf('Equispaced ($\\Lambda_{%d}$ = %.1f)', N, max(Lambda_equi)));
plot(x_fine, Lambda_leg, 'Color', TEAL, 'LineWidth', 1.5, ...
     'DisplayName', sprintf('Legendre ($\\Lambda_{%d}$ = %.2f)', N, max(Lambda_leg)));
plot(x_fine, Lambda_cheb, 'Color', SKY, 'LineWidth', 1.5, ...
     'DisplayName', sprintf('Chebyshev ($\\Lambda_{%d}$ = %.2f)', N, max(Lambda_cheb)));

yline(1, '--', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5, 'HandleVisibility', 'off');

xlabel('$x$', 'Interpreter', 'latex');
ylabel('$\Lambda_N(x)$', 'Interpreter', 'latex');
title(sprintf('Lebesgue Functions ($N = %d$)', N), 'FontSize', 11);
xlim([-1, 1]);
ylim([0, 35]);
legend('Location', 'north', 'Interpreter', 'latex');

text(-0.6, 25, 'Peaks at boundaries', 'FontSize', 9, 'Color', [0.5, 0.5, 0.5]);

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

% Right panel: Growth of Lebesgue constants
subplot(1, 2, 2);
hold on; box on;

N_values = 2:30;

Lambda_equi_vals = zeros(size(N_values));
Lambda_cheb_vals = zeros(size(N_values));
Lambda_leg_vals = zeros(size(N_values));

for i = 1:length(N_values)
    N = N_values(i);

    % Equispaced
    x_equi = linspace(-1, 1, N+1)';
    Lambda_equi_vals(i) = max(lebesgue_function(x_equi, x_fine));

    % Chebyshev
    j = 0:N;
    x_cheb = cos(j * pi / N)';
    Lambda_cheb_vals(i) = max(lebesgue_function(x_cheb, x_fine));

    % Legendre
    x_leg = legendre_nodes(N);
    Lambda_leg_vals(i) = max(lebesgue_function(x_leg, x_fine));
end

semilogy(N_values, Lambda_equi_vals, 'o-', 'Color', CORAL, 'LineWidth', 1.5, ...
         'MarkerSize', 4, 'DisplayName', 'Equispaced');
semilogy(N_values, Lambda_leg_vals, 's-', 'Color', TEAL, 'LineWidth', 1.5, ...
         'MarkerSize', 4, 'DisplayName', 'Legendre');
semilogy(N_values, Lambda_cheb_vals, '^-', 'Color', SKY, 'LineWidth', 1.5, ...
         'MarkerSize', 4, 'DisplayName', 'Chebyshev');

% Asymptotic curve for Chebyshev
N_asy = linspace(5, 30, 100);
Lambda_cheb_asy = (2/pi) * log(N_asy) + 0.6;
h_asy = semilogy(N_asy, Lambda_cheb_asy, '--', 'Color', SKY, 'LineWidth', 1, ...
         'DisplayName', '$(2/\pi)\ln N$');
h_asy.Color(4) = 0.5;  % Set alpha transparency

xlabel('$N$ (polynomial degree)', 'Interpreter', 'latex');
ylabel('Lebesgue constant $\Lambda_N$', 'Interpreter', 'latex');
title('Growth of Lebesgue Constants', 'FontSize', 11);
xlim([0, 32]);
legend('Location', 'northwest', 'Interpreter', 'latex');

% Add growth rate labels
text(25, 1e6, '$O(2^N)$', 'Interpreter', 'latex', 'FontSize', 10, 'Color', CORAL);
text(25, 4, '$O(\sqrt{N})$', 'Interpreter', 'latex', 'FontSize', 10, 'Color', TEAL);
text(25, 2.5, '$O(\ln N)$', 'Interpreter', 'latex', 'FontSize', 10, 'Color', SKY);

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

%% Save figure
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

exportgraphics(fig, output_file, 'ContentType', 'vector');
fprintf('Figure saved to: %s\n', output_file);

png_file = strrep(output_file, '.pdf', '.png');
exportgraphics(fig, png_file, 'Resolution', 300);

%% Print summary
fprintf('\nLebesgue Constants Comparison:\n');
fprintf('%s\n', repmat('-', 1, 65));
fprintf('%4s %15s %15s %15s\n', 'N', 'Equispaced', 'Legendre', 'Chebyshev');
fprintf('%s\n', repmat('-', 1, 65));
for i = 1:5:length(N_values)
    N = N_values(i);
    fprintf('%4d %15.2f %15.4f %15.4f\n', N, ...
            Lambda_equi_vals(i), Lambda_leg_vals(i), Lambda_cheb_vals(i));
end
fprintf('%s\n', repmat('-', 1, 65));

%% Helper functions
function Lambda = lebesgue_function(x_nodes, x_eval)
    N = length(x_nodes) - 1;
    Lambda = zeros(size(x_eval));

    for k = 1:N+1
        L_k = ones(size(x_eval));
        for j = 1:N+1
            if j ~= k
                L_k = L_k .* (x_eval - x_nodes(j)) / (x_nodes(k) - x_nodes(j));
            end
        end
        Lambda = Lambda + abs(L_k);
    end
end

function x = legendre_nodes(N)
    % Legendre-Gauss-Lobatto nodes
    % LGL nodes are: x_0 = -1, x_N = 1, and interior nodes are zeros of P'_N(x)
    if N == 0
        x = 0;
        return;
    end
    if N == 1
        x = [-1; 1];
        return;
    end

    % Start with Chebyshev-Gauss-Lobatto nodes as initial guess
    j = 0:N;
    x = -cos(j * pi / N)';  % Note: sorted from -1 to 1

    % Newton iteration for interior nodes only (endpoints are fixed at ±1)
    x(1) = -1;
    x(N+1) = 1;

    for iter = 1:20
        [~, dP] = legendre_poly(N, x(2:N));
        [~, ~, d2P] = legendre_poly_with_d2(N, x(2:N));

        % Newton step: find zeros of P'_N(x)
        dx = dP ./ d2P;
        x(2:N) = x(2:N) - dx;

        if max(abs(dx)) < 1e-15
            break;
        end
    end
    x = sort(x);
end

function [P, dP] = legendre_poly(N, x)
    % Evaluate Legendre polynomial P_N and its derivative at x
    P_prev = ones(size(x));
    P_curr = x;

    if N == 0
        P = P_prev;
        dP = zeros(size(x));
        return;
    end
    if N == 1
        P = P_curr;
        dP = ones(size(x));
        return;
    end

    for n = 1:N-1
        P_next = ((2*n+1)*x.*P_curr - n*P_prev) / (n+1);
        P_prev = P_curr;
        P_curr = P_next;
    end
    P = P_curr;

    % Derivative: P'_N(x) = N*(x*P_N - P_{N-1}) / (x^2 - 1)
    dP = N * (x.*P - P_prev) ./ (x.^2 - 1 + eps);
    % Handle endpoints
    dP(abs(x - 1) < 1e-10) = N*(N+1)/2;
    dP(abs(x + 1) < 1e-10) = (-1)^(N+1) * N*(N+1)/2;
end

function [P, dP, d2P] = legendre_poly_with_d2(N, x)
    % Evaluate Legendre polynomial P_N, its first and second derivatives at x
    % Uses three-term recurrence for P_n and derivative relations

    P_prev = ones(size(x));
    P_curr = x;
    dP_prev = zeros(size(x));
    dP_curr = ones(size(x));
    d2P_prev = zeros(size(x));
    d2P_curr = zeros(size(x));

    if N == 0
        P = P_prev;
        dP = dP_prev;
        d2P = d2P_prev;
        return;
    end
    if N == 1
        P = P_curr;
        dP = dP_curr;
        d2P = d2P_curr;
        return;
    end

    for n = 1:N-1
        % Recurrence for P_{n+1}
        P_next = ((2*n+1)*x.*P_curr - n*P_prev) / (n+1);

        % Recurrence for P'_{n+1} (differentiate the recurrence)
        dP_next = ((2*n+1)*(P_curr + x.*dP_curr) - n*dP_prev) / (n+1);

        % Recurrence for P''_{n+1}
        d2P_next = ((2*n+1)*(2*dP_curr + x.*d2P_curr) - n*d2P_prev) / (n+1);

        P_prev = P_curr;
        P_curr = P_next;
        dP_prev = dP_curr;
        dP_curr = dP_next;
        d2P_prev = d2P_curr;
        d2P_curr = d2P_next;
    end

    P = P_curr;
    dP = dP_curr;
    d2P = d2P_curr;
end

File: codes/julia/ch04/lebesgue_functions.jl
# lebesgue_functions.jl
#
# Visualizes and compares Lebesgue functions for different node distributions.
#
# The Lebesgue function is:
#     Lambda_N(x) = sum_{k=0}^{N} |L_k(x)|
#
# where L_k(x) are the Lagrange basis polynomials. The Lebesgue constant is:
#     Lambda_N = max_{x in [-1,1]} Lambda_N(x)
#
# The Lebesgue constant bounds the interpolation error:
#     ||f - p_N||_inf <= (1 + Lambda_N) E_N(f)
#
# Asymptotic growth rates:
#     - Chebyshev: Lambda_N = (2/pi) ln(N) + O(1)
#     - Legendre:  Lambda_N ~ O(sqrt(N))
#     - Equispaced: Lambda_N ~ 2^N / (N ln N) (exponential!)
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LaTeXStrings
using Polynomials

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N_FINE = 1000

# Book color scheme
const NAVY  = colorant"#142D6E"
const SKY   = colorant"#7896D2"
const CORAL = colorant"#E74C3C"
const TEAL  = colorant"#16A085"

# Output path
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch04", "julia")

# -----------------------------------------------------------------------------
# Node generation
# -----------------------------------------------------------------------------
equispaced_nodes(N) = collect(range(-1, 1, length=N+1))
chebyshev_nodes(N) = [cos(j * π / N) for j in 0:N]

"""
    legendre_nodes(N)

Legendre-Gauss-Lobatto nodes on [-1, 1].
These are +/-1 plus the N-1 zeros of P'_N(x).
"""
function legendre_nodes(N)
    N == 0 && return [0.0]
    N == 1 && return [-1.0, 1.0]

    # Build Legendre polynomial P_N in the Legendre basis
    # Coefficients: all zeros except the N-th is 1
    coeffs = zeros(N + 1)
    coeffs[N + 1] = 1.0

    # Convert to standard polynomial using recurrence relation
    # P_0(x) = 1, P_1(x) = x, (n+1)P_{n+1}(x) = (2n+1)x P_n(x) - n P_{n-1}(x)
    P_prev = [1.0]       # P_0
    P_curr = [0.0, 1.0]  # P_1

    for n in 1:N-1
        P_next = ((2n + 1) .* [0; P_curr] .- n .* [P_prev; zeros(2)]) ./ (n + 1)
        # Pad P_prev for next iteration
        P_prev = P_curr
        P_curr = P_next
    end

    # P_curr is now P_N(x) in monomial basis
    # Compute derivative
    P_N_poly = Polynomial(P_curr)
    P_N_deriv = derivative(P_N_poly)

    # Find roots of P'_N
    interior = roots(P_N_deriv)
    interior = real.(filter(r -> abs(imag(r)) < 1e-10, interior))

    return sort(vcat([-1.0], interior, [1.0]))
end

# -----------------------------------------------------------------------------
# Lebesgue function computation
# -----------------------------------------------------------------------------
function lagrange_basis(x_nodes, k, x_eval)
    n = length(x_nodes)
    L_k = ones(length(x_eval))
    for j in 1:n
        if j != k
            L_k .*= (x_eval .- x_nodes[j]) ./ (x_nodes[k] - x_nodes[j])
        end
    end
    return L_k
end

function lebesgue_function(x_nodes, x_eval)
    N = length(x_nodes) - 1
    Lambda = zeros(length(x_eval))
    for k in 1:N+1
        Lambda .+= abs.(lagrange_basis(x_nodes, k, x_eval))
    end
    return Lambda
end

function lebesgue_constant(x_nodes; n_eval=2000)
    x_eval = collect(range(-1, 1, length=n_eval))
    Lambda = lebesgue_function(x_nodes, x_eval)
    return maximum(Lambda)
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Fine grid for plotting
    x_fine = collect(range(-1, 1, length=N_FINE))

    # Create figure with two panels
    fig = Figure(size = (1100, 450))

    # ==========================================================================
    # Left panel: Lebesgue functions for N = 10
    # ==========================================================================
    N_plot = 10

    x_equi = equispaced_nodes(N_plot)
    x_cheb = chebyshev_nodes(N_plot)
    x_leg = legendre_nodes(N_plot)

    Lambda_equi = lebesgue_function(x_equi, x_fine)
    Lambda_cheb = lebesgue_function(x_cheb, x_fine)
    Lambda_leg = lebesgue_function(x_leg, x_fine)

    ax1 = Axis(fig[1, 1],
               xlabel = L"x", ylabel = L"\Lambda_N(x)",
               title = "Lebesgue Functions (N = $N_plot)",
               limits = ((-1, 1), (0, 35)))

    lines!(ax1, x_fine, Lambda_equi, color = CORAL, linewidth = 1.5,
           label = latexstring("Equispaced (\\Lambda_{$N_plot} = $(round(maximum(Lambda_equi), digits=1)))"))
    lines!(ax1, x_fine, Lambda_leg, color = TEAL, linewidth = 1.5,
           label = latexstring("Legendre (\\Lambda_{$N_plot} = $(round(maximum(Lambda_leg), digits=2)))"))
    lines!(ax1, x_fine, Lambda_cheb, color = SKY, linewidth = 1.5,
           label = latexstring("Chebyshev (\\Lambda_{$N_plot} = $(round(maximum(Lambda_cheb), digits=2)))"))

    hlines!(ax1, 1, color = colorant"#888888", linewidth = 0.5, linestyle = :dash)

    # Annotation
    text!(ax1, -0.6, 25.0, text = "Peaks at boundaries\nfor equispaced nodes",
          fontsize = 9, color = :gray)

    axislegend(ax1, position = :ct, framevisible = false, labelsize = 9)

    # ==========================================================================
    # Right panel: Lebesgue constant growth with N
    # ==========================================================================
    N_values = 2:30

    Lambda_equi_values = [lebesgue_constant(equispaced_nodes(n)) for n in N_values]
    Lambda_cheb_values = [lebesgue_constant(chebyshev_nodes(n)) for n in N_values]
    Lambda_leg_values = [lebesgue_constant(legendre_nodes(n)) for n in N_values]

    ax2 = Axis(fig[1, 2],
               xlabel = L"N" * " (polynomial degree)",
               ylabel = L"\Lambda_N" * " (Lebesgue constant)",
               title = "Growth of Lebesgue Constants",
               limits = ((0, 32), nothing),
               yscale = log10)

    scatterlines!(ax2, collect(N_values), Lambda_equi_values,
                  color = CORAL, linewidth = 1.5, markersize = 6, marker = :circle,
                  label = "Equispaced")
    scatterlines!(ax2, collect(N_values), Lambda_leg_values,
                  color = TEAL, linewidth = 1.5, markersize = 6, marker = :rect,
                  label = "Legendre")
    scatterlines!(ax2, collect(N_values), Lambda_cheb_values,
                  color = SKY, linewidth = 1.5, markersize = 6, marker = :utriangle,
                  label = "Chebyshev")

    # Plot asymptotic curve for Chebyshev
    N_asy = range(5, 30, length=100)
    Lambda_cheb_asy = (2/π) .* log.(N_asy) .+ 0.6
    lines!(ax2, collect(N_asy), Lambda_cheb_asy, linestyle = :dash, color = SKY,
           linewidth = 1, alpha = 0.5, label = L"(2/\pi)\ln N")

    # Growth rate annotations
    text!(ax2, 25.0, 1e6, text = L"O(2^N)", fontsize = 10, color = CORAL)
    text!(ax2, 25.0, 4.0, text = L"O(\sqrt{N})", fontsize = 10, color = TEAL)
    text!(ax2, 25.0, 2.5, text = L"O(\ln N)", fontsize = 10, color = SKY)

    axislegend(ax2, position = :lt, framevisible = false, labelsize = 9)

    # Save figure
    mkpath(OUTPUT_DIR)
    save(joinpath(OUTPUT_DIR, "lebesgue_functions.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "lebesgue_functions.png"), fig, px_per_unit = 4)

    println("Figure saved to: ", joinpath(OUTPUT_DIR, "lebesgue_functions.pdf"))

    # Print summary table
    println("\nLebesgue Constants Comparison:")
    println("-" ^ 65)
    println("$(rpad("N", 4)) $(lpad("Equispaced", 15)) $(lpad("Legendre", 15)) $(lpad("Chebyshev", 15))")
    println("-" ^ 65)
    for i in 1:5:length(N_values)
        n = N_values[i]
        println("$(lpad(n, 4)) $(lpad(round(Lambda_equi_values[i], digits=2), 15)) $(lpad(round(Lambda_leg_values[i], digits=4), 15)) $(lpad(round(Lambda_cheb_values[i], digits=4), 15))")
    end
    println("-" ^ 65)

    println("\nAsymptotic Growth Rates:")
    println("  Chebyshev:  Lambda_N ~ (2/pi) ln(N) + O(1)")
    println("  Legendre:   Lambda_N ~ O(sqrt(N))")
    println("  Equispaced: Lambda_N ~ 2^(N+1) / (eN ln N)")

    return fig
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
